lib/connections/quic_listen.go

```diff
@@ -0,0 +1,408 @@
+	if err != nil {
+		t.mut.Lock()
+		t.err = err
+		t.mut.Unlock()
+		l.Infoln("Listen (BEP/quic):", err)
+		return
+	}
+
+	defer listener.Close()
+	defer stunConn.Close()
+	defer quicConn.Close()
+	defer registry.Unregister(t.uri.Scheme, quicConn.(net.PacketConn))
+	defer filterConn.Close()
+	defer packetConn.Close()
+
+	l.Infof("QUIC listener (%v) starting", quicConn.LocalAddr())
+	defer l.Infof("QUIC listener (%v) shutting down", quicConn.LocalAddr())
+
+	go t.stunRenewal(stunConn, writeTrackingConn.GetLastWrite)
+
+	// Accept is forever, so handle stops externally.
+	go func() {
+		select {
+		case <-t.stop:
+			_ = listener.Close()
+		}
+	}()
+
+	for {
+		// Blocks forever, see https://github.com/lucas-clemente/quic-go/issues/1915
+		session, err := listener.Accept()
+
+		select {
+		case <-t.stop:
+			if err == nil {
+				_ = session.Close()
+			}
+			return
+		default:
+		}
+		if err != nil {
+			if err, ok := err.(net.Error); !ok || !err.Timeout() {
+				l.Warnln("Listen (BEP/quic): Accepting connection:", err)
+			}
+			continue
+		}
+
+		l.Debugln("connect from", session.RemoteAddr())
+
+		// Accept blocks forever, give it 10s to do it's thing.
+		ok := make(chan struct{})
+		go func() {
+			select {
+			case <-ok:
+				return
+			case <-time.After(10 * time.Second):
+				l.Debugln("timed out waiting for AcceptStream on", session.RemoteAddr())
+				_ = session.Close()
+			}
+		}()
+
+		stream, err := session.AcceptStream()
+		close(ok)
+		if err != nil {
+			l.Debugln("failed to accept stream from", session.RemoteAddr(), err.Error())
+			_ = session.Close()
+			continue
+		}
+
+		t.conns <- internalConn{&quicTlsConn{session, stream}, connTypeQUICServer, quicPriority}
+	}
+}
+
+func (t *quicListener) Stop() {
+	close(t.stop)
+}
+
+func (t *quicListener) URI() *url.URL {
+	return t.uri
+}
+
+func (t *quicListener) WANAddresses() []*url.URL {
+	uris := t.LANAddresses()
+	t.mut.RLock()
+	if t.address != nil {
+		uris = append(uris, t.address)
+	}
+	t.mut.RUnlock()
+	return uris
+}
+
+func (t *quicListener) LANAddresses() []*url.URL {
+	return []*url.URL{t.uri}
+}
+
+func (t *quicListener) Error() error {
+	t.mut.RLock()
+	err := t.err
+	t.mut.RUnlock()
+	return err
+}
+
+func (t *quicListener) String() string {
+	return t.uri.String()
+}
+
+func (t *quicListener) Factory() listenerFactory {
+	return t.factory
+}
+
+func (t *quicListener) NATType() string {
+	v := t.nat.Load().(stun.NATType)
+	if v == stun.NATUnknown || v == stun.NATError {
+		return "unknown"
+	}
+	return v.String()
+}
+
+func (t *quicListener) stunRenewal(listener net.PacketConn, getLastWriteTime func() time.Time) {
+	for {
+	disabled:
+		if t.stunDisabled() {
+			time.Sleep(time.Second)
+			continue
+		}
+
+		l.Debugf("Starting stun for %s", t.uri)
+
+		t.nat.Store(stun.NATUnknown)
+		t.mut.Lock()
+		t.address = nil
+		t.mut.Unlock()
+
+		for _, addr := range t.cfg.StunServers() {
+			if !t.runStunForServer(listener, addr, getLastWriteTime) {
```

Comment:
Please add some commentary here about `runStunForServer` not actually returning unless the server fails, as that part is quite deep in the call stack and it took me a little while to figure out that this doesn't buggily run one stun transaction against every server in sequence